File: Structural/Proxy/proxy.py
class Subject(ABC):
    @abstractmethod
    def request(self): pass

class RealSubject(Subject):
    def request(self): print("RealSubject: Handling request.")

class Proxy(Subject):
    def __init__(self, real_subject: RealSubject):
        self._real_subject = real_subject

    def request(self):
        print("Proxy: Logging access before calling RealSubject.")
        self._real_subject.request()

File: Structural/Proxy/proxy.cs
public interface ISubject { void Request(); }

class RealSubject : ISubject {
    public void Request() => Console.WriteLine("RealSubject: Handling Request.");
}

class Proxy : ISubject {
    private RealSubject _realSubject;
    public void Request() {
        if (_realSubject == null) _realSubject = new RealSubject();
        Console.WriteLine("Proxy: Checking access before firing real request.");
        _realSubject.Request();
    }
}